File: dynamic_programming/house_robber.cpp
// Bottom-up implementation
class Solution {
public:
    // TC: O( sizeof( nums ) )
    // SC: O( 1 )
    int rob(vector<int>& nums)
    {
        if( nums.size() == 1 )
            return nums.front();
        
        if( nums.size() == 2 )
            return std::max( nums[ 0 ], nums[ 1 ] );
        
        nums[ 2 ] += nums[ 0 ];
        int sz = nums.size();
        for( int i = 3; i < sz; ++i )
            nums[ i ] += std::max( nums[ i - 2 ], nums[ i - 3 ] );
        
        return std::max( nums[ sz - 1 ], nums[ sz - 2 ] );
    }
};

// Top-down implementation
class Solution {
public:
    // TC: O( sizeof( nums ) )
    // SC: O( sizeof( nums ) )
    int rob(vector<int>& nums)
    {
        if( nums.size() == 1 )
            return nums.front();
        
        if( nums.size() == 2 )
            return std::max( nums[ 0 ], nums[ 1 ] );
        
        nums[ 2 ] += nums[ 0 ];
        check_houses( nums, 3 );
        return std::max( nums[ nums.size() - 1 ], nums[ nums.size() - 2 ] );
    }
    
    void check_houses( std::vector<int>& nums, int i ) {
        if( i >= nums.size() )
            return;
        
        nums[ i ] += std::max( nums[ i - 2 ], nums[ i - 3 ] );
        check_houses( nums, i + 1 );
    }
};

File: dynamic_programming/house_robber.rs
// Bottom-up implementation
impl Solution {
    // TC: O( sizeof( nums ) )
    // SC: O( 1 )
    pub fn rob(mut nums: Vec<i32>) -> i32 {
        if nums.len() == 1 {
            return nums[ 0 ];
        }
        
        if nums.len() == 2 {
            return if nums[ 0 ] > nums[ 1 ] { nums[ 0 ] } else { nums[ 1 ] };
        }
        
        nums[ 2 ] += nums[ 0 ];
        for i in 3..nums.len() {
            nums[ i ] += std::cmp::max( nums[ i - 2 ], nums[ i - 3 ] );
        }
        
        let sz = nums.len();
        return std::cmp::max( nums[ sz - 1 ], nums[ sz - 2 ] );
    }
}

// Top-down implementation
impl Solution {
    pub fn rob(mut nums: Vec<i32>) -> i32 {
        if nums.len() == 1 {
            return nums[ 0 ];
        }
        
        if nums.len() == 2 {
            return std::cmp::max( nums[ 0 ], nums[ 1 ] );
        }
        
        nums[ 2 ] += nums[ 0 ];
        Self::check_houses( &mut nums, 3 );
        return std::cmp::max( nums[ nums.len() - 1 ], nums[ nums.len() - 2 ] );
    }
    
    pub fn check_houses( nums: &mut Vec<i32>, i: usize ) {
        if i >= nums.len() {
            return;
        }
        
        nums[ i ] += std::cmp::max( nums[ i - 2 ], nums[ i - 3 ] );
        Self::check_houses( nums, i + 1 );
    } 
}
